### PR TITLE
Webpack 2 is here!

### PR DESCRIPTION
**Pull request by choonkending**
# Reason
#743 - spike webpack 2.0

but mainly due to @lapinskicho 's https://github.com/reactGo/reactGo/pull/806 (thanks again!!)

# High Level Changes
- Use #806 's changes
- Remove stylelint as dependency
- Replace NoErrorsPlugin with NoEmitOnErrorsPlugin
♻️ 
⚠️ The changes here are opinionated, i.e. there are good reasons to keep configurations within a single file. However, I believe that splitting a large configuration file into smaller manageable pieces will be worth its value in the long run as it will make webpack more manageable. I think this outweighs the cons (having extra mental overhead in finding modules). I spent quite some time refactoring and coming up with meaningful names + folder structure, I _really_ hope this will make things more maintainable!

- Move webpack configurations into `/webpack`
- Create `webpack/rules/` which define
    - javascript (babel-loader) - I thought hard about this one. Babel is an implementation detail rather than _what_ we want, which is a loader for javascript with support for jsx + es2015.
    - image (url-loader
    - css (css-loader + postcss-loader + friends)
- Split out plugins to `webpack/plugins`
- Split out externals to `webpack/externals`
- Extract PATHS to `webpack/paths`

**Important upgrades**
- webpack - 2.2.1 🎉 🎉 

🔧 
- Fix issue raised in https://github.com/reactGo/reactGo/pull/806#discussion_r96241135 regarding production server render

# Questions
1. This is a huge change. How do I review it?
Good question. Go through each logical commit. Each commit should be a small change enough to review in nice bite-sized chunks.

Thank you!

### CODE REVIEW

**Comment by choonkending** [comment]
> I removed --debug from "build:dev" in package.json, if you want to keep it you may want to add it back.

Thanks for that! I'll add that back in for dev mode!

**Comment by choonkending** [comment]
Merging in an hour!

File: app/tests/karma.conf.js [hunk 1]
```diff
@@ -35,24 +34,24 @@ module.exports = function(config) {
...
         loaders: [
           {
             test: /\.js$|\.jsx$/,
-            loader: 'babel',
+            loader: 'babel-loader',
             // Reason why we put this here instead of babelrc
             // https://github.com/gaearon/react-transform-hmr/issues/5#issuecomment-142313637
-            query: {
-              'presets': ['es2015', 'react', 'stage-0'],
-              'plugins': [
```
Comment (psimyn) [inline app/tests/karma.conf.js]: Is this hmr thing still an issue?

**Comment by psimyn** [comment]
images

**Pull request merged**